src/hooks/useWatchHistory.ts

```diff
@@ -0,0 +1,123 @@
+import { MutableRefObject, useCallback, useEffect, useLayoutEffect, useMemo, useRef } from 'react';
+
+import type { JWPlayer } from '#types/jwplayer';
+import type { PlaylistItem } from '#types/playlist';
+import useEventCallback from '#src/hooks/useEventCallback';
+import { saveItem } from '#src/stores/WatchHistoryController';
+import { useConfigStore } from '#src/stores/ConfigStore';
+import { useWatchHistoryStore } from '#src/stores/WatchHistoryStore';
+import { VideoProgressMinMax } from '#src/config';
+
+type QueuedProgress = {
+  item: PlaylistItem;
+  progress: number;
+};
+
+/**
+ * The `useWatchHistory` hook has two responsibilities.
+ *
+ * 1. Determine the `startTime` of the given `item` (return value)
+ * 2. Save the watch progress of the current `item` to the user profile
+ *
+ * __The problem:__
+ *
+ * There are multiple events that trigger a save. This results in duplicate (unnecessary) saves and API calls. Another
+ * problem is that some events are triggered when the `item` to update has changed. For example, when clicking a media
+ * item in the "Related shelf". This causes the wrong item to be saved in the watch history.
+ *
+ * __The solution:__
+ *
+ * This hook listens to the player time update event and queues a watch history entry with the current progress and
+ * item. When this needs to be saved, the queue is used to look up the last progress and item and save it when
+ * necessary. The queue is then cleared to prevent duplicate saves and API calls.
+ *
+ * @param player
+ * @param item
+ */
+export const useWatchHistory = (player: JWPlayer | undefined, item: PlaylistItem) => {
+  const queuedWatchProgress = useRef<QueuedProgress | null>(null) as MutableRefObject<QueuedProgress | null>;
+
+  // config
+  const { features } = useConfigStore((s) => s.config);
+  const continueWatchingList = features?.continueWatchingList;
+  const watchHistoryEnabled = !!continueWatchingList;
+
+  // watch History
+  const watchHistoryItem = useWatchHistoryStore((state) => (!!item && watchHistoryEnabled ? state.getItem(item) : undefined));
+
+  // update the queued watch progress on each time update event
+  const handleTimeUpdate = useEventCallback(() => {
```

Comment:
Do player and item need to be dependencies for this callback? Or are they global / never change?

Reply:
I use `useEventCallback` here to prevent the callback from updating when the `item` changes because it will re-trigger the `useEffect` that (un)binds the event listeners. We had problems with that before regarding certain events being missed.

The `useEventCallback` doesn't take any dependencies and will always use the latest state/variables.